File: go/shield/signatures.go
package shield

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"time"

	"golang.org/x/crypto/pbkdf2"
)

var (
	// ErrLamportKeyUsed indicates the Lamport key was already used.
	ErrLamportKeyUsed = errors.New("shield: Lamport key already used")
	// ErrInvalidSignature indicates signature verification failed.
	ErrInvalidSignature = errors.New("shield: invalid signature")
)

// SymmetricSignature provides HMAC-based signatures.
type SymmetricSignature struct {
	signingKey      [KeySize]byte
	VerificationKey [KeySize]byte
}

// NewSymmetricSignature creates a signature from a signing key.
func NewSymmetricSignature(signingKey []byte) (*SymmetricSignature, error) {
	if len(signingKey) != KeySize {
		return nil, ErrInvalidKeySize
	}

	ss := &SymmetricSignature{}
	copy(ss.signingKey[:], signingKey)

	// Derive verification key
	h := sha256.New()
	h.Write([]byte("verify:"))
	h.Write(signingKey)
	copy(ss.VerificationKey[:], h.Sum(nil))

	return ss, nil
}

// GenerateSymmetricSignature generates a new random signature identity.
func GenerateSymmetricSignature() (*SymmetricSignature, error) {
	key := make([]byte, KeySize)
	if _, err := rand.Read(key); err != nil {
		return nil, err
	}
	return NewSymmetricSignature(key)
}

// SymmetricSignatureFromPassword derives signature from password and identity.
func SymmetricSignatureFromPassword(password, identity string) *SymmetricSignature {
	salt := sha256.Sum256([]byte("sign:" + identity))
	key := pbkdf2.Key([]byte(password), salt[:], 100000, KeySize, sha256.New)

	ss := &SymmetricSignature{}
	copy(ss.signingKey[:], key)

	h := sha256.New()
	h.Write([]byte("verify:"))
	h.Write(key)
	copy(ss.VerificationKey[:], h.Sum(nil))

	return ss
}

// Sign signs a message.
func (ss *SymmetricSignature) Sign(message []byte, includeTimestamp bool) []byte {
	if includeTimestamp {
		timestamp := time.Now().Unix()
		tsBytes := make([]byte, 8)
		binary.LittleEndian.PutUint64(tsBytes, uint64(timestamp))

		sigData := append(tsBytes, message...)
		mac := hmac.New(sha256.New, ss.signingKey[:])
		mac.Write(sigData)
		sig := mac.Sum(nil)

		return append(tsBytes, sig...)
	}

	mac := hmac.New(sha256.New, ss.signingKey[:])
	mac.Write(message)
	return mac.Sum(nil)
}

// Verify verifies a signature.
func (ss *SymmetricSignature) Verify(message, signature []byte, verificationKey []byte, maxAge int64) bool {
	if subtle.ConstantTimeCompare(verificationKey, ss.VerificationKey[:]) != 1 {
		return false
	}

	if len(signature) == 40 {
		// Timestamped signature
		timestamp := binary.LittleEndian.Uint64(signature[:8])
		sig := signature[8:]

		if maxAge > 0 {
			now := time.Now().Unix()
			diff := now - int64(timestamp)
			if diff < 0 {
				diff = -diff
			}
			if diff > maxAge {
				return false
			}
		}

		// Create new slice to avoid mutating signature
		sigData := make([]byte, 8+len(message))
		copy(sigData[:8], signature[:8])
		copy(sigData[8:], message)

		mac := hmac.New(sha256.New, ss.signingKey[:])
		mac.Write(sigData)
		expected := mac.Sum(nil)

		return subtle.ConstantTimeCompare(sig, expected) == 1
	}

	if len(signature) == 32 {
		mac := hmac.New(sha256.New, ss.signingKey[:])
		mac.Write(message)
		expected := mac.Sum(nil)
		return subtle.ConstantTimeCompare(signature, expected) == 1
	}

	return false
}

// Fingerprint returns the key fingerprint.
func (ss *SymmetricSignature) Fingerprint() string {
	h := sha256.Sum256(ss.VerificationKey[:])
	return hex.EncodeToString(h[:8])
}

// LamportSignature provides one-time post-quantum signatures.
type LamportSignature struct {
	privateKey [][2][KeySize]byte
	PublicKey  []byte
	used       bool
}

// GenerateLamportSignature generates a new Lamport key pair.
func GenerateLamportSignature() (*LamportSignature, error) {
	const bits = 256
	ls := &LamportSignature{
		privateKey: make([][2][KeySize]byte, bits),
		PublicKey:  make([]byte, bits*64),
	}

	for i := 0; i < bits; i++ {
		if _, err := rand.Read(ls.privateKey[i][0][:]); err != nil {
			return nil, err
		}
		if _, err := rand.Read(ls.privateKey[i][1][:]); err != nil {
			return nil, err
		}

		h0 := sha256.Sum256(ls.privateKey[i][0][:])
		h1 := sha256.Sum256(ls.privateKey[i][1][:])

		copy(ls.PublicKey[i*64:i*64+32], h0[:])
		copy(ls.PublicKey[i*64+32:i*64+64], h1[:])
	}

	return ls, nil
}

// Sign signs a message (ONE TIME ONLY).
func (ls *LamportSignature) Sign(message []byte) ([]byte, error) {
	if ls.used {
		return nil, ErrLamportKeyUsed
	}
	ls.used = true

	msgHash := sha256.Sum256(message)
	signature := make([]byte, 256*32)

	for i := 0; i < 256; i++ {
		byteIdx := i / 8
		bitIdx := i % 8
		bit := (msgHash[byteIdx] >> bitIdx) & 1

		if bit == 1 {
			copy(signature[i*32:(i+1)*32], ls.privateKey[i][1][:])
		} else {
			copy(signature[i*32:(i+1)*32], ls.privateKey[i][0][:])
		}
	}

	return signature, nil
}

// VerifyLamport verifies a Lamport signature.
func VerifyLamport(message, signature, publicKey []byte) bool {
	if len(signature) != 256*32 || len(publicKey) != 256*64 {
		return false
	}

	msgHash := sha256.Sum256(message)

	for i := 0; i < 256; i++ {
		byteIdx := i / 8
		bitIdx := i % 8
		bit := (msgHash[byteIdx] >> bitIdx) & 1

		revealed := signature[i*32 : (i+1)*32]
		hashed := sha256.Sum256(revealed)

		var expected []byte
		if bit == 1 {
			expected = publicKey[i*64+32 : i*64+64]
		} else {
			expected = publicKey[i*64 : i*64+32]
		}

		if subtle.ConstantTimeCompare(hashed[:], expected) != 1 {
			return false
		}
	}

	return true
}

// IsUsed returns whether the key has been used.
func (ls *LamportSignature) IsUsed() bool {
	return ls.used
}

// Fingerprint returns the public key fingerprint.
func (ls *LamportSignature) Fingerprint() string {
	h := sha256.Sum256(ls.PublicKey)
	return hex.EncodeToString(h[:8])
}

File: csharp/Shield/Signatures.cs
using System;
using System.Security.Cryptography;
using System.Text;

namespace Guard8.Shield
{
    /// <summary>
    /// SymmetricSignature provides HMAC-based signatures.
    /// </summary>
    public class SymmetricSignature : IDisposable
    {
        private readonly byte[] _signingKey;
        private readonly byte[] _verificationKey;
        private bool _disposed;

        public SymmetricSignature(byte[] signingKey)
        {
            if (signingKey.Length != Shield.KeySize)
                throw new ArgumentException("Invalid key size", nameof(signingKey));

            _signingKey = new byte[Shield.KeySize];
            Array.Copy(signingKey, _signingKey, Shield.KeySize);

            // Derive verification key
            byte[] data = new byte[7 + Shield.KeySize];
            Encoding.UTF8.GetBytes("verify:").CopyTo(data, 0);
            Array.Copy(signingKey, 0, data, 7, Shield.KeySize);
            _verificationKey = Shield.Sha256(data);
        }

        public static SymmetricSignature Generate()
        {
            return new SymmetricSignature(Shield.RandomBytes(Shield.KeySize));
        }

        public static SymmetricSignature FromPassword(string password, string identity)
        {
            byte[] salt = Shield.Sha256(Encoding.UTF8.GetBytes("sign:" + identity));
            byte[] key = Shield.Pbkdf2(password, salt, Shield.Iterations, Shield.KeySize);
            return new SymmetricSignature(key);
        }

        public byte[] Sign(byte[] message, bool includeTimestamp = false)
        {
            if (includeTimestamp)
            {
                long timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
                byte[] tsBytes = BitConverter.GetBytes(timestamp);

                byte[] sigData = new byte[8 + message.Length];
                Array.Copy(tsBytes, 0, sigData, 0, 8);
                Array.Copy(message, 0, sigData, 8, message.Length);

                byte[] sig = Shield.HmacSha256(_signingKey, sigData);

                byte[] result = new byte[40];
                Array.Copy(tsBytes, 0, result, 0, 8);
                Array.Copy(sig, 0, result, 8, 32);
                return result;
            }

            return Shield.HmacSha256(_signingKey, message);
        }

        public bool Verify(byte[] message, byte[] signature, byte[] verificationKey, long maxAge = 0)
        {
            if (!Shield.ConstantTimeEquals(verificationKey, _verificationKey, Shield.KeySize))
                return false;

            if (signature.Length == 40)
            {
                long timestamp = BitConverter.ToInt64(signature, 0);

                if (maxAge > 0)
                {
                    long now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
                    long diff = Math.Abs(now - timestamp);
                    if (diff > maxAge)
                        return false;
                }

                byte[] sigData = new byte[8 + message.Length];
                Array.Copy(signature, 0, sigData, 0, 8);
                Array.Copy(message, 0, sigData, 8, message.Length);

                byte[] expected = Shield.HmacSha256(_signingKey, sigData);

                byte[] receivedSig = new byte[32];
                Array.Copy(signature, 8, receivedSig, 0, 32);
                return Shield.ConstantTimeEquals(receivedSig, expected, 32);
            }

            if (signature.Length == 32)
            {
                byte[] expected = Shield.HmacSha256(_signingKey, message);
                return Shield.ConstantTimeEquals(signature, expected, 32);
            }

            return false;
        }

        public byte[] VerificationKey
        {
            get
            {
                byte[] copy = new byte[_verificationKey.Length];
                Array.Copy(_verificationKey, copy, _verificationKey.Length);
                return copy;
            }
        }

        public string Fingerprint()
        {
            byte[] hash = Shield.Sha256(_verificationKey);
            return BitConverter.ToString(hash, 0, 8).Replace("-", "").ToLower();
        }

        public void Dispose()
        {
            if (!_disposed)
            {
                Shield.SecureWipe(_signingKey);
                Shield.SecureWipe(_verificationKey);
                _disposed = true;
            }
        }
    }

    /// <summary>
    /// LamportSignature provides one-time post-quantum signatures.
    /// </summary>
    public class LamportSignature : IDisposable
    {
        private readonly byte[,,] _privateKey;  // [256, 2, 32]
        private readonly byte[] _publicKey;     // [256 * 64]
        private bool _used;
        private bool _disposed;

        private LamportSignature()
        {
            _privateKey = new byte[256, 2, Shield.KeySize];
            _publicKey = new byte[256 * 64];
            _used = false;
        }

        public static LamportSignature Generate()
        {
            var ls = new LamportSignature();

            for (int i = 0; i < 256; i++)
            {
                byte[] key0 = Shield.RandomBytes(Shield.KeySize);
                byte[] key1 = Shield.RandomBytes(Shield.KeySize);

                for (int j = 0; j < Shield.KeySize; j++)
                {
                    ls._privateKey[i, 0, j] = key0[j];
                    ls._privateKey[i, 1, j] = key1[j];
                }

                byte[] h0 = Shield.Sha256(key0);
                byte[] h1 = Shield.Sha256(key1);

                Array.Copy(h0, 0, ls._publicKey, i * 64, 32);
                Array.Copy(h1, 0, ls._publicKey, i * 64 + 32, 32);
            }

            return ls;
        }

        public byte[] Sign(byte[] message)
        {
            if (_used)
                throw new InvalidOperationException("Lamport key already used");
            _used = true;

            byte[] msgHash = Shield.Sha256(message);
            byte[] signature = new byte[256 * 32];

            for (int i = 0; i < 256; i++)
            {
                int byteIdx = i / 8;
                int bitIdx = i % 8;
                int bit = (msgHash[byteIdx] >> bitIdx) & 1;

                for (int j = 0; j < 32; j++)
                    signature[i * 32 + j] = _privateKey[i, bit, j];
            }

            return signature;
        }

        public static bool Verify(byte[] message, byte[] signature, byte[] publicKey)
        {
            if (signature.Length != 256 * 32 || publicKey.Length != 256 * 64)
                return false;

            byte[] msgHash = Shield.Sha256(message);

            for (int i = 0; i < 256; i++)
            {
                int byteIdx = i / 8;
                int bitIdx = i % 8;
                int bit = (msgHash[byteIdx] >> bitIdx) & 1;

                byte[] revealed = new byte[32];
                Array.Copy(signature, i * 32, revealed, 0, 32);
                byte[] hashed = Shield.Sha256(revealed);

                byte[] expected = new byte[32];
                if (bit == 1)
                    Array.Copy(publicKey, i * 64 + 32, expected, 0, 32);
                else
                    Array.Copy(publicKey, i * 64, expected, 0, 32);

                if (!Shield.ConstantTimeEquals(hashed, expected, 32))
                    return false;
            }

            return true;
        }

        public bool IsUsed => _used;

        public byte[] PublicKey
        {
            get
            {
                byte[] copy = new byte[_publicKey.Length];
                Array.Copy(_publicKey, copy, _publicKey.Length);
                return copy;
            }
        }

        public string Fingerprint()
        {
            byte[] hash = Shield.Sha256(_publicKey);
            return BitConverter.ToString(hash, 0, 8).Replace("-", "").ToLower();
        }

        public void Dispose()
        {
            if (!_disposed)
            {
                for (int i = 0; i < 256; i++)
                {
                    for (int j = 0; j < 32; j++)
                    {
                        _privateKey[i, 0, j] = 0;
                        _privateKey[i, 1, j] = 0;
                    }
                }
                _disposed = true;
            }
        }
    }
}
